src/utils/calculator.ts

```diff
@@ -0,0 +1,67 @@
+import { GatewayOracle } from './gateway_oracle';
+import type { ArweaveOracle } from './arweave_oracle';
+import regression, { DataPoint } from 'regression';
+
+const SET = 'SET';
+
+export class Calculator {
+	private static readonly sampleByteVolumes = [
+		102_400, // 100 KiB
+		104_857_600, // 100 MiB
+		107_374_182_400 // 10 GiB
+	];
+	private sampleWinstonValues: number[] = [];
+	private regressionInstance?: regression.Result;
+
+	private hasData = false;
+	private isSyncing = false;
+	private setupPromise?: Promise<void>;
+
+	constructor(doNotSetup = false, private oracle: ArweaveOracle = new GatewayOracle()) {
+		if (!doNotSetup) {
+			this.setup();
+		}
+	}
+
+	private async setup(): Promise<void> {
+		if (this.isSyncing) {
+			return this.setupPromise;
+		}
+		if (!this.hasData) {
+			this.isSyncing = true;
+			const workerPromise = this.setupWorker();
+			this.setupPromise = workerPromise;
+			await workerPromise;
+			this.isSyncing = false;
+			this.setupPromise = undefined;
+		}
+		this.hasData = true;
+	}
+
+	private async setupWorker(): Promise<void> {
+		await this.fetchData();
+		this.runTheLinearRegression();
+	}
+
+	private async fetchData(): Promise<void> {
+		const winstonSamplesQuery = Calculator.sampleByteVolumes.map((sampleByteCount) =>
+			this.oracle.getWinstonPriceForByteCount(sampleByteCount)
+		);
+		const winstonSamplesResult = await Promise.all(winstonSamplesQuery);
+		this.sampleWinstonValues = winstonSamplesResult;
+	}
+
+	private runTheLinearRegression(): void {
+		const dataPoints: DataPoint[] = Calculator.sampleByteVolumes.map(
+			(bytesCount, index) => [bytesCount, this.sampleWinstonValues[index]] as DataPoint
+		);
+		this.regressionInstance = regression.linear(dataPoints);
+	}
+
+	public async getPriceForBytes(byteCount: number): Promise<number | undefined> {
```

Comment:
We should be clear here that this returns as Winston in the variable name.

Also as a user of this class, this is a hard function to use. Every time we use this function it should be after startup, after the one and only async call has resolved. Having to await the synchronous return of regressionInstance.predict causes the result to fall on to the next event loop.

Seeing how the class has ended up, I would rather store the `regressionInstance` in state so we can synchronously call `predict` on it as needed.

Reply:
Given the way I've refactored this, I think in most cases you will simply execute on the current event loop because no awaits will take place when a regression is cached. Let me know if you believe otherwise.